File: gpt-3.5-ft-large/stage2/data_deviceptr_clause.f90
#ifndef T1
!T1:data deviceptr clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        REAL(8),DIMENSION(10):: b
        REAL(8),DIMENSION(10):: c
        REAL(8),DIMENSION(10), DEVICE :: a
        INTEGER,DEVICE::dev
        REAL(8):: temp
        INTEGER:: random_seed
        LOGICAL,DIMENSION(1)::devtest
        REAL(8),DIMENSION(10):: host_a
        REAL(8),DIMENSION(10):: host_c
        INTEGER:: x

        !Initilization on host
        devtest(1) = .TRUE.
        !$acc enter data copyin(devtest(1:1))
        !$acc parallel present(devtest(1:1))
          dev = 0
          dev = omp_get_initial_device()
          IF (dev .eq. acc_get_device_num(acc_get_device_type())) then
            devtest(1) = .FALSE.
          END IF
        !$acc end parallel
        !$acc exit data delete(devtest(1:1))

        SEED('data') 
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        CALL RANDOM_NUMBER(b)
        c = 0
        host_a = a
        host_c = c

        !Calculation on device
        !$acc data deviceptr(a)
          !$acc parallel loop
          DO i = 1, 10
            a(i) = a(i) + b(i)
          END DO
        !$acc end data
        
        !Calculation on host
        DO i = 1, 10
          host_a(i) = host_a(i) + b(i)
        END DO

        !Copy back to host from device.
        c = host_c
        a = host_a

        !Verification
        DO i = 1, 10
          IF (ABS(a(i) - c(i)) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO


        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END
#endif

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
#ifndef T1
        LOGICAL :: test1
#endif
        failed = .FALSE.
        failcode = 0
#ifndef T1
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
#endif
        CALL EXIT (failcode)
      END PROGRAM

File: Phind-CodeLlama-34B-v2/stage2/context_template_full/data_deviceptr_clause.cpp
#include <iostream>
#include <cstdlib>
#include <cassert>
#include "acc_testsuite.h"

#ifndef T1
/*T1:data deviceptr clause,V:2.0-2.7*/
int test1(){
    int err = 0;
    srand(SEED);

    int N = 100;
    int *host_array = new int[N];
    int *device_array;

    for (int i = 0; i < N; i++) {
        host_array[i] = rand() % 100;
    }

    #pragma acc enter data copyin(host_array[0:N])
    #pragma acc data deviceptr(device_array)
    {
        #pragma acc host_data use_device(device_array)
        {
            device_array = host_array;
        }

        #pragma acc parallel present(device_array[0:N])
        {
            for (int i = 0; i < N; i++) {
                device_array[i] *= 2;
            }
        }

        #pragma acc update host(host_array[0:N])
    }
    #pragma acc exit data delete(device_array)

    for (int i = 0; i < N; i++) {
        if (host_array[i] != (rand() % 100) * 2) {
            err = 1;
            break;
        }
    }

    delete[] host_array;

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}